### ISSUE TITLE
Multiple liquid tag helpers not supported

### ISSUE THREAD

**Issue by arkadiuszwojcik** [OrchardCMS/OrchardCore#8012]
Today I was experimenting with liqid helpers and tried to make use of `input`. Liquid script I had was:

`{% helper "input", type: "text", value: "some_val", for: "some_field" %}`

I could not make this code work, what was rendered was empty input control: `<input>`

What I found out debuging Orchard source code was that: `LiquidTagHelperFactory.GetActivator("input" ...)` is not returning `InputTagHelper` as I would expect but `InputIsDisabledTagHelper` which is registred earlier. Both tag helpers use same name: `input`.

Edit:

Maybe `LiquidTagHelperFactory.GetActivator("input" ...)` should be prepared to return multiple tag helpers with same name? Such filters should be then sorted by `TagHelper.Order `property.

**Comment by jtkech** [comment]
@arkadiuszwojcik 

Good catch, we already have the same issue with our `cache` tag helper that we renamed `dynamic-cache` as i remember, so that it doesn't conflict with the mvc `cache` one that is only selected on the `cache` tag name.

Here easy to fix, we only need to add the `asp-is-disabled` attribute to the `HtmlTargetElement`

    [HtmlTargetElement("input", Attributes = ForAndIsDisabledAttributeName)]
    public class InputIsDisabledTagHelper : InputTagHelper
    {
        private const string ForAndIsDisabledAttributeName = "asp-for, asp-is-disabled";

        [HtmlAttributeName("asp-is-disabled")]
        public bool IsDisabled { set; get; }

Do you want to suggest a PR for this as you found the issue, otherwise no problem i will do it soon ;)

**Comment by arkadiuszwojcik** [comment]
@jtkech I will do PR today with this fix. Besides that, is it reasonable to create story that will request multiple TagHelpers support for liquid or is it too much effort?

**Comment by deanmarcussen** [comment]
@arkadiuszwojcik @jtkech 

We've seen a little issue with this tag helper before.

I don't think it should be inheriting from `InputTagHelper` - there is no need for it too, and it is why it can get triggered when it shouldn't. Just inherited from `TagHelper` should be enough for it. 

The changes you suggest to the attributes are good as well JT

**Comment by jtkech** [comment]
I'm doing some tests

Yes, through liquid we don't execute a set of tag helpers

So, with the above issue we can get the mvc input working through liquid, but then we can't get the `InputIsDisabledTagHelper` working, but it's better to have the mvc one working

But because `InputIsDisabledTagHelper` inherits from the `InputTagHelper`, maybe we could just also process the base one, but then it may be processed twice, i'm doing some tests

> is it reasonable to create story that will request multiple TagHelpers support for liquid or is it too much effort?

Maybe yes but why not ;)

@deanmarcussen just saw your message, let me do some tests ;)

**Comment by deanmarcussen** [comment]
@jtkech cool. 

For infos. 

We fixed another issue a while back by stopping the call to base (it was triggering twice), but I think the right fix, was to not inherit.

**Comment by jtkech** [comment]
@arkadiuszwojcik @deanmarcussen 

Okay in razor all matching helpers are executed which is intended when doing complementary works and in the order of the overridable `Order` property. The problem we had with our `cache` tag helper was that it is not complementary with the mvc one and we needed to not execute the mvc one, so we renamed our tag helper.

> We fixed another issue a while back by stopping the call to base (it was triggering twice)

Yes indeed if it was only doing a complementary work, and because the matched base one is still executing by itself

> but I think the right fix, was to not inherit.

Not sure if we don't call the base processing, but anyway at least it is useless to inherit
So yes, you're right here better that `InputIsDisabledTagHelper` just inherits from `TagHelper`

> The changes you suggest to the attributes are good as well JT

Yes anyway, it prevents our `InputIsDisabledTagHelper` tag, as i saw, to be executed for each input even without the `is-disabled` attribute.

So here, after adding a matching attribute, we can now fully use the mvc input under liquid, the last goal is to be still able to use `InputIsDisabledTagHelper` under liquid when we use the `is-disabled` attribute. Just tried by just inheriting from `TagHelper`, under liquid `InputIsDisabledTagHelper` is executed, i understand why, when derived from `InputTagHelper` it was inheriting an `Order` property of -1000, now it has a default `Order` of 0

But because the mvc one is not executed, yes the input is disabled but in the rendered html tag we loose the related generated attributes. So here the missing thing is that under liquid we don't execute all matching helpers, which may be good in some case ;) but not here.

 Maybe worth to do, maybe not as here **i don't see the benefit of the `InputIsDisabledTagHelper` as it just adds a `disabled="disabled"` attribute that is easy to add explicitly as is.**

**Comment by arkadiuszwojcik** [comment]
@jtkech @deanmarcussen 

Let me do quick summary of potential fixes:

1. `InputIsDisabledTagHelper ` should inherit `InputTagHelper` and have additional `'asp-is-disabled'` constraint. This will work for Liquid as only one tag helper will be used anyway. But such solution will break Razor tag helper because it will not only match `InputIsDisabledTagHelper` but also `InputTagHelper` (which is less restrictive), so logic for `input` will be executd twice. Not good option - Razor pages with this tag helper will be affected.
2. `InputIsDisabledTagHelper`  should inherit `TagHelper` and have additional `'asp-is-disabled'` constraint. This will work partialy, as regular input logic provided by `InputTagHelper` will be not executed. Not good option - `InputIsDisabledTagHelper`  will be still useless in Liquid and this will probably brake things if it is used in Liquid contexts already.
3. Remove `InputIsDisabledTagHelper`. Will require some code refactory work. It will fix current broken logic in Liquid by not overriding `InputTagHelper.` Also location of `InputIsDisabledTagHelper` is a bit odd, taking in consideration its siblings tags.
4. Implement #8019, this is most complex work but is most elegant solution because it reflect Razor strategy of dealing with TagHelpers. All matching TagHelpers will have to be executed by liquid and then `InputIsDisabledTagHelper` will be complementary to `InputTagHelper.`

Is it all?

For me looks like to best way to go is to do 3rd option (then this issue can be close) and later/optionaly 4th option.

**Comment by jtkech** [comment]
@arkadiuszwojcik 

> 1. `InputIsDisabledTagHelper ` should inherit `InputTagHelper` and have additional `'asp-is-disabled'` constraint. This will work for Liquid as only one tag helper will be used anyway. But such solution will break Razor tag helper because it will not only match `InputIsDisabledTagHelper` but also `InputTagHelper` (which is less restrictive), so logic for `input` will be executd twice. Not good option - Razor pages with this tag helper will be affected.

No, currently for razor both tag helpers are executed and each only once as `InputIsDisabledTagHelper` doesn't call the `Process()` of its base class. But anyway not good to inherit from `InputTagHelper`, see below

> 2. `InputIsDisabledTagHelper`  should inherit `TagHelper` and have additional `'asp-is-disabled'` constraint. This will work partialy, as regular input logic provided by `InputTagHelper` will be not executed. Not good option - `InputIsDisabledTagHelper` will be still useless in Liquid and this will probably brake things if it is used in Liquid contexts already.

That's okay for razor as the `InputTagHelper` is still selected and processed, and if `InputIsDisabledTagHelper` is also selected by an additional `asp-is-disabled` it will be also processed to add the html `disabled` attribute, this without calling the `Process()` of its base class, as before.

And that's okay for liquid if they were used without `asp-is-disabled` as it will work better by really using the mvc helper, and i think it is very rare that they were used with `asp-is-disabled`, anyway it was not fully working, only the disabled html attribute was added but other helper attributes as `value` were lost.

Another advantage of the additional constraint is that before in razor the `InputIsDisabledTagHelper` was always executed, even without an `asp-is-disabled` attribute

Here another advantage is that before `InputTagHelper` was inheriting an `Order` property of -1000, then it would be 0 which is better and ensures that the only one helper that would be used in liquid would be the mvc one.

> 3. Remove `InputIsDisabledTagHelper`. Will require some code refactory work. It will fix current broken logic in Liquid by not overriding `InputTagHelper.` Also location of `InputIsDisabledTagHelper` is a bit odd, taking in consideration its siblings tags.

That's an option, **and i agree about the location ;)**

> 4. Implement #8019, this is most complex work but is most elegant solution because it reflect Razor strategy of dealing with TagHelpers. All matching TagHelpers will have to be executed by liquid and then `InputIsDisabledTagHelper` will be complementary to `InputTagHelper.`

I agree, more complex especially since they moved this part from mvc to the razor language

> For me looks like to best way to go is to do 3rd option (then this issue can be close) and later/optionaly 4th option.

Yes to not do 4., yes that 3. is an option, **but i think that for now it is better to do 2.**
